flowkit_jwt_generator/flowkit_jwt_generator/jwt.py

```diff
@@ -0,0 +1,213 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import datetime
+import os
+import uuid
+from json import JSONEncoder
+from typing import Dict, List, Union, Callable
+from datetime import timedelta
+import click
+import jwt
+import pytest
+import requests
+
+
+permissions_types = {"run": True, "poll": True, "get_result": True}
+
+
+def generate_token(
+    username: str,
+    secret: str,
+    lifetime: datetime.timedelta,
+    claims: Dict[str, Union[Dict[str, bool], List[str]]],
+) -> str:
+    """
+
+    Parameters
+    ----------
+    username : str
+        Username for the token
+    secret : str
+        Shared secret to sign the token with
+    lifetime : datetime.timedelta
+        Lifetime from now of the token
+    claims : dict
+        Dictionary of claims the token will grant
+
+    Returns
+    -------
+    str
+        Encoded token
+
+    """
+    now = datetime.datetime.utcnow()
+    token_data = dict(
+        iat=now,
+        nbf=now,
+        jti=str(uuid.uuid4()),
+        user_claims=claims,
+        identity=username,
+        exp=now + lifetime,
+    )
+    return jwt.encode(
+        payload=token_data, key=secret, algorithm="HS256", json_encoder=JSONEncoder
+    ).decode("utf-8")
+
+
+@pytest.fixture
+def access_token_builder() -> Callable:
+    """
+    Fixture which builds short-life access tokens.
+
+    Returns
+    -------
+    function
+        Function which returns a token encoding the specified claims.
+    """
+
+    secret = os.getenv("JWT_SECRET_KEY")
+    if secret is None:
+        raise EnvironmentError("JWT_SECRET_KEY environment variable not set.")
+
+    def token_maker(claims):
+        return generate_token(
+            username="test",
+            secret=secret,
+            lifetime=timedelta(seconds=90),
+            claims=claims,
+        )
+
+    return token_maker
+
+
+def get_all_claims_from_flowapi(
+    flowapi_url: str
+) -> Dict[str, Union[Dict[str, bool], List[str]]]:
+    """
+    Retrieve all the queries types available on an instance of flowapi and
+    generate a claims dictionary which grants total access to them.
+
+    Parameters
+    ----------
+    flowapi_url : str
+        URL of the flowapi instance
+
+    Returns
+    -------
+    dict
+        Claims dictionary
+
+    """
+    query_kinds = []
+    aggregation_types = set()
+    spec = requests.get(f"{flowapi_url}/api/0/spec/openapi.json").json()
+    for q_class, q_spec in spec["components"]["schemas"].items():
+        try:
+            query_kinds.append(q_spec["properties"]["query_kind"]["enum"][0])
+            aggregation_types.update(q_spec["properties"]["aggregation_unit"]["enum"])
+        except KeyError:
```

Comment:
There are quite a few possibilities why a `KeyError` may be thrown here (missing `"properties"`, missing `"query_kind"`, missing `"aggregation_unit"`, missing `"enum"). Are we sure this only every happens for the reason we expect, and we're not accidentally end up missing some claims? Probably fine, just checking.

Reply:
Anything missing any of properties, query kind, or the enum key isn't a query kind, anything missing an aggregation_unit enum doesn't itself know about age types, so I think this is fine?